src/selectors/entities/channels.js

```diff
@@ -321,3 +331,185 @@ export const canManageChannelMembers = createSelector(
     getLicense,
     canManageMembers
 );
+
+export const getDirectChannelIds = createIdsSelector(
+    getDirectChannelsSet,
+    (directIds) => {
+        return Array.from(directIds);
+    }
+);
+
+export const getChannelIdsInCurrentTeam = createIdsSelector(
+    getCurrentTeamId,
+    getChannelsInTeam,
+    (currentTeamId, channelsInTeam) => {
+        return Array.from(channelsInTeam[currentTeamId] || []);
+    }
+);
+
+export const getChannelIdsForCurrentTeam = createIdsSelector(
+    getChannelIdsInCurrentTeam,
+    getDirectChannelIds,
+    (channels, direct) => {
+        return [...channels, ...direct];
+    }
+);
+
+export const getUnreadChannelIds = createIdsSelector(
+    getAllChannels,
+    getMyChannelMemberships,
+    getChannelIdsForCurrentTeam,
+    (channels, members, teamChannelIds) => {
+        return teamChannelIds.filter((id) => {
+            const c = channels[id];
+            const m = members[id];
+
+            if (c && m) {
+                if (m.notify_props && m.notify_props.mark_unread !== 'mention' && c.total_msg_count - m.msg_count > 0) {
+                    return true;
+                }
+            }
+            return false;
+        });
+    }
+);
+
+export const getSortedUnreadChannelIds = createIdsSelector(
+    getCurrentUser,
+    getUsers,
+    getAllChannels,
+    getMyChannelMemberships,
+    getUnreadChannelIds,
+    getTeammateNameDisplaySetting,
+    (currentUser, profiles, channels, members, unreadIds, settings) => {
```

Comment:
this will perform a shallowEquality with the arguments, if one of them is different it will re-run the function but if the result of the function is shallow equal to the last result then it will return the memoized result so it is strict equal

Reply:
Ok, so the savings isn't in not re-running the function it's in returning the memoized result so the components don't re-render if the result didn't change. Got it, thanks.